craft_parts/sources/sources.py

```diff
@@ -71,16 +71,113 @@
 cases you want to refer to the documentation for the specific plugin.
 """
 
-from typing import Dict, Type
+import os
+import re
+from typing import TYPE_CHECKING, Dict, Optional, Type
 
+from craft_parts.dirs import ProjectDirs
+
+from . import errors
 from .base import SourceHandler
 from .local_source import LocalSource
 from .tar_source import TarSource
 
+if TYPE_CHECKING:
+    from craft_parts.parts import Part
+
 SourceHandlerType = Type[SourceHandler]
 
 
 _source_handler: Dict[str, SourceHandlerType] = {
     "local": LocalSource,
     "tar": TarSource,
 }
+
+
+def get_source_handler(
+    application_name: str,
+    part: "Part",
+    dirs: ProjectDirs,
+) -> Optional[SourceHandler]:
+    """Return the appropriate handler for the given source.
+
+    :param application_name: The name of the application using Craft Parts.
+    :param part: The part to get a source handler for.
+    :param dirs: The project's work directories.
+    """
+    if not dirs:
+        dirs = ProjectDirs()
+
+    source_handler = None
+    if part.spec.source:
+        handler_class = _get_source_handler_class(
+            part.spec.source,
+            source_type=part.spec.source_type,
+        )
+        source_handler = handler_class(
+            application_name=application_name,
```

Comment:
refresh my memory on why the application name is important here

Reply:
It's used to compose the path to the downloaded file cache.

Reply:
Ah, maybe this can be global, or in other words, scoped to craft-parts and not specific applications.